### PR TITLE
http: changing some Connection* interfaces to be OptRef

### PR DESCRIPTION
**Pull request by alyssawilk**
Risk Level: low
Testing: existing tests
Docs Changes: n/a
Release Notes: n/a
part of https://github.com/envoyproxy/envoy/issues/10455

### CODE REVIEW

File: source/common/router/router.cc [hunk 1]
```diff
@@ -50,11 +50,11 @@ constexpr char NumInternalRedirectsFilterStateName[] = "num_internal_redirects";
 uint32_t getLength(const Buffer::Instance* instance) { return instance ? instance->length() : 0; }
 
 bool schemeIsHttp(const Http::RequestHeaderMap& downstream_headers,
-                  const Network::Connection& connection) {
+                  OptRef<const Network::Connection> connection) {
   if (downstream_headers.getSchemeValue() == Http::Headers::get().SchemeValues.Http) {
     return true;
   }
-  if (!connection.ssl()) {
+  if (connection && !connection->ssl()) {
```
Comment (jmarantz) [inline source/common/router/router.cc:57]: nit: we've mostly been using `connection.has_value()` here, though if this compiles I suppose this must be correct.

File: source/common/http/filter_manager.cc [hunk 2]
```diff
@@ -1335,7 +1337,9 @@ void FilterManager::setBufferLimit(uint32_t new_limit) {
 }
 
 void FilterManager::contextOnContinue(ScopeTrackedObjectStack& tracked_object_stack) {
-  tracked_object_stack.add(connection_);
+  if (connection_.has_value()) {
```
Comment (jmarantz) [inline source/common/http/filter_manager.cc:1340]: this appears to be a semantic change. Were we previously adding potentially null-pointers to the stack?

File: source/common/router/upstream_request.cc [hunk 3]
```diff
@@ -213,16 +213,18 @@ void UpstreamRequest::decodeTrailers(Http::ResponseTrailerMapPtr&& trailers) {
 void UpstreamRequest::dumpState(std::ostream& os, int indent_level) const {
   const char* spaces = spacesForLevel(indent_level);
   os << spaces << "UpstreamRequest " << this << "\n";
-  const auto addressProvider = connection().connectionInfoProviderSharedPtr();
+  if (connection()) {
```
Comment (jmarantz) [inline source/common/router/upstream_request.cc:216]: again this appears to be a semantic change; was this previously a potential crasher?

File: source/extensions/common/wasm/context.cc [hunk 4]
```diff
@@ -1108,9 +1108,9 @@ StreamInfo::StreamInfo* Context::getRequestStreamInfo() const {
 
 const Network::Connection* Context::getConnection() const {
   if (encoder_callbacks_) {
-    return encoder_callbacks_->connection();
+    return encoder_callbacks_->connection().ptr();
```
Comment (jmarantz) [inline source/extensions/common/wasm/context.cc:1111]: in a follow-up will you change getConnection() to also return an OptRef?

File: source/extensions/upstreams/http/tcp/upstream_request.cc [hunk 5]
```diff
@@ -52,9 +52,9 @@ Envoy::Http::Status TcpUpstream::encodeHeaders(const Envoy::Http::RequestHeaderM
   if (route_entry->connectConfig().has_value()) {
     Buffer::OwnedImpl data;
     auto& connect_config = route_entry->connectConfig().value();
-    if (connect_config.has_proxy_protocol_config()) {
+    if (connect_config.has_proxy_protocol_config() && upstream_request_->connection()) {
```
Comment (jmarantz) [inline source/extensions/upstreams/http/tcp/upstream_request.cc:55]: nit (your choice): use connection().has_value()

**Review by jmarantz — COMMENTED** [review]
lgtm basically with a couple of nits I'll leave up to you, and a couple of small questions on semantics.

File: source/common/http/filter_manager.cc (hunk 2 above)
Comment (alyssawilk) [reply source/common/http/filter_manager.cc:1340]: are you asking if I'm fixing a crashing bug in public?  No :-P  the interface allowed for null pointers, but the implementation with a hard-coded Network::Connection& did not.  For upstream filters, there may actually not be a downstream connection in the case of async requests, so for that we have to handle this case.

File: source/extensions/common/wasm/context.cc (hunk 4 above)
Comment (alyssawilk) [reply source/extensions/common/wasm/context.cc:1111]: maybe?  :-)

**Review by jmarantz — APPROVED** [review]

**Pull request merged**